book/src/proposals/leader-duplicate-block-slashing.md

```diff
@@ -0,0 +1,224 @@
+
+5) `Dead Slots` keyspace: `(slot, blockhash)`
+Validators need to specify which version of a slot is dead. For instance if one
+version of a slot doesn't pass entry verification, but another version does,
+then the two versions must be distinguishable. 
+
+An issue to note here is that if some version of a slot violates correctness
+before the slot is finished, then the validator does not know what the ending
+blockhash is and thus cannot store the slot as dead. In these cases we can drop
+the entire slot and wait for repair. More details in the `Replay Failures` 
+section below.
+
+### Repairing Multiple versions of the Same Slot
+Repair is augmented with a blockhash. The various types of repairs:
+
+`Shred(slot, blockhash, index)` - Ask for a specific shred in a slot
+with version `blockhash`. 
+
+Notes:
+
+1) Setting `blockhash` to `Hash::default()` means this validator assumes there
+is only one version of this slot in the cluster and will acccept any shred
+for `slot` and `index`. However, if a validator gets a mixed bag of such
+shreds that do not chain, they will then drop the entire `slot`, and only
+repair `slot` if some later child slot chains back to `slot` (this will be 
+initiated by `Orphan` repairs below). 
+
+2) Repair responses need to be tied to a particular blockhash. However, 
+because repair responses are limited to the size of a shred, we cannot
+include the blockhash in the repair response. This means we need to include 
+some repair "cookie" in the request + response that maps some number of bits
+to a particular `blockhash`. This breaks down if there are more versions of 
+this slot than can be tracked by the number of bits allocated to the cookie.
+
+
+`Orphan(slot, first_tick_hash, num_hashes)` - Ask for the last shred
+of the parent of the child slot, where the child has slot equal to `slot`. 
+This last shred must contain the last tick `T_p` in that parent slot such that
+hashing `T_p` `num_hashes` times is equal to `first_tick_hash`.
+
+Notes:
+
+1) The response will include the last shred of the parent slot. From this
+last shred the validator should be able to get the last tick and the 
+merkle root, which can be used to compute the `blockhash`. This `blockhash`
+is then used to make requests of the form `Shred(slot, blockhash, index)`
+to repair the rest of this slot.
+
+2) Also requires a cookie similar to the `Shred(slot, blockhash, index)` 
+requests above.
+
+### Chaining Multiple Versions of the Same Slot
+When there are multiple versions for a slot `A`, a natural question that arises
+is how validators know which version of slot `A` some descendant slot `B`
+chains to. Shreds currently specify only a `parent` slot, but not which version 
+of that parent. The approach to figure out chaining behavior is then:
+
+1) Wait for all the shreds for the first entry `E_B` of slot `B` to arrive 
+(Implementation can make sure first shred `S_B` always contains only a tick 
+to avoid waiting for multiple shreds).
```

Comment:
this option would delay ingest for 25% of the slot...

Reply:
we can add a "first shred in slot" and have that shred carry the parent blockhash

Reply:
another option: give up 4-8 bytes of shred space to carry that many bits of parent blockhash.  this would solve  the repair cookie issues, too.

Reply:
@rob-solana, that first option works, marking and having the first shred carry the parent blockhash. Why does it need the first tick hash though?

Reply:
I was proposing alternatives.  I think the 4-8 bytes of parent blockhash would be sufficient, ignore others.

Reply:
> @sagar-solana, @rob-solana, @aeyakovenko, @ryoqun The next area of concern is surrounding the status cache and how it's packaged into snapshots.

This sounds like a real mess.
 
> 1. The root in blocktree, and thus the root included in the snapshot will now have to carry which version of that slot was rooted (specified by blockhash)
> 2. Need to update the status cache to handle multiple versions of the same slot, meaning status cache will have to support a key by `(slot, blockhash)`
> 3. Before we detect a conflicting version of a slot/when the leader is building the block in banking stage, the identifier for a slot in the status cache will be `Hash::default()`. (Recall the blockhash is known for any versions of the slot after a conflict is detected because we repair those versoins of the slot by blockhash)

Only during catchup or repair will the blockhash be known a-priori, so in the normal case the status cache will have to re-index at the end of each block, too?

Status cache is organized by fork (i.e. is chained in memory via bank construction), right?  If it were, would there be a need to key by blockhash?

Reply:
yeah its a huge mess 😛 

Status cache is a global store shared across all the banks. At a high level, it maps from siganature -> Vec<(Slot, Status)>. When we look up a signature, we pass a vector of ancestors and check if any of those ancestors contain that signature. 

Because now we can have many different versions of a single ancestor, we will have to index by blockhash.

Reply:
ok, now I remember...

instead of using slots as ancestors in status cache, what if banks assigned themselves a random number and we used that as ancestry in statuscache?

signature->Hashmap<BankID, Status> ?